File: Java/Hussain Sir Java/Fundamentals/Language/Inheritance/InterfaceTest2.java
interface Consumer extends AutoCloseable{

	int LIMIT = 10;

	void apply(int action);

	default void applyAll(int first, int last){
		for(int act = first; act <= last; ++act)
			apply(act);
	}
}

class ResourceConsumer implements Consumer{

	private String id;

	public ResourceConsumer(String name){
		id = name;
		System.out.printf("%s resource acquired%n", id);
	}

	public void apply(int action){
		if(id != null)
			System.out.printf("%s resource consumed using action<%d>%n", id, action);
	}

	public void close(){
		System.out.printf("%s resource released%n", id);
		id = null;
	}

}

class InterfaceTest2{

	private static void run(String op){
		try(var b = new ResourceConsumer("Second")){
			int n = Integer.parseInt(op);
			b.applyAll(1, n);
		}
	}

	public static void main(String[] args){
		var a = new ResourceConsumer("First");
		a.apply(1);
		a.close();
		try{
			run(args[0]);
		}catch(Throwable t){}
	}
}

File: Fundamentals/Language/Inheritance/interfacetest2.cs
using System;

interface IConsumer : IDisposable
{
	void Apply(int action);
}

static class Consumers
{
	//extension-method: a method of a static class which can be called as
	//an instance method of its first 'this' qualified parameter type
	public static void ApplyAll(this IConsumer that, int first, int last)
	{
		for(int action = first; action <= last; ++action)
			that.Apply(action);
	}
}

class ResourceConsumer : IConsumer
{
	private string id;

	public ResourceConsumer(string name)
	{
		id = name;
		Console.WriteLine($"{id} resource acquired.");
	}

	//explicit implementation of Apply method defined in IConsumer interface
	void IConsumer.Apply(int action)
	{
		Console.WriteLine($"{id} resource consumed using action<{action}>");
	}

	void IDisposable.Dispose()
	{
		Console.WriteLine($"{id} resource released.");
	}
}

static class Program
{
	private static void Run(string up)
	{
		using(IConsumer b = new ResourceConsumer("Second"))
		{
			b.ApplyAll(1, int.Parse(up)); //Consumers.ApplyAll(b, 1, ...);
		}
		
	}
	
	public static void Main(string[] args)
	{
		IConsumer a = new ResourceConsumer("First");
		a.Apply(1);
		a.Dispose();

		try
		{
			Run(args[0]);
		}
		catch {}
	}
}
